### PR TITLE
feat: add a scheduler MVP

### PR DESCRIPTION
**Pull request by lucklove**

### CODE REVIEW

File: requirements.txt
[the remark this answers is not in the document]
Comment (lucklove) [reply requirements.txt:1]: No, it's not necessary. Developers can choose their favorite tools.

File: model.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    """
+    Application models app. An app has many versions but only one of them are active.
+    Each time the app is called, the active_version's configration is used to build
+    the DAG, and then produce the interaction record.
+    """
+
+    __tablename__ = "applications"
+
+    id = Column(String(36), primary_key=True, nullable=False)
+    name = Column(String(255), nullable=False)
```
[the remark this answers is not in the document]
Comment (lucklove) [reply model.py:37]: Currently, `name` is not searchable. Maybe we will support search by name in the future, then we can index it.

File: database.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            app (Application): The application object to be added.
+
+        Returns:
+            None
+        """
+        with Session(self.engine) as session:
+            session.add(app)
+            session.commit()
+
+    def get_application(self, app_id: str) -> Application:
```
[the remark this answers is not in the document]
Comment (lucklove) [reply database.py:47]: Yes, the deletion mark only apply on list apis. So it's possible for a user to call a deleted application.

File: api/application.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+                parent_id=version.parent_id,
+                created_at=created_at,
+                updated_at=created_at,
+                configuration=version.configuration.dict(),
+            )
+        )
+        return VersionCreateResponse(id=_id)
+
+    @router.delete("/applications/{application_id}/versions/{version_id}")
+    def delete_app_version(
```
[the remark this answers is not in the document]
Comment (lucklove) [reply api/application.py:388]: Maybe a check is required, active_version can not be deleted.

[the remark this answers is not in the document]
Comment (lucklove) [reply api/application.py:388]: A check has been added.

File: database.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (freedeaths) [reply database.py:47]: It is a little weird that an active version is deleted but still available.

File: blocks/invoke.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    timeout: int = 300,
+) -> str:
+    env = Env()
+    env.read_env()
+
+    db = Database(create_engine(env.str("DATABASE_URL")))
+    invoker = AsyncInvoker(db)
+
+    interaction_id = invoker.invoke(app_id, input)
+    while timeout > 0:
```
Comment (freedeaths) [inline blocks/invoke.py:236]: the value of `timeout` doesn't count down.

**Review by freedeaths — CHANGES_REQUESTED** [review]

File: blocks/invoke.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+            app_id (str): The ID of the application to invoke.
+            input (Union[str, dict, list]): The input data for the application,
+                which can be a string, dictionary, or list.
+
+        Returns:
+            str: The ID of the interaction created for this invocation.
+        """
+        app = self.database.get_application(app_id)
+        if not app:
+            raise ApplicationNotFound(app_id)
```
Comment (freedeaths) [inline blocks/invoke.py:137]: not imported

File: blocks/invoke.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        if not app.active_version:
+            NoActiveVersion(app_id)
```
Comment (freedeaths) [inline blocks/invoke.py:139]: the same as line 137

File: blocks/invoke.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        version = self.database.get_version(app.active_version)
+        if not version:
+            raise VersionnNotFound(app.active_version)
```
Comment (freedeaths) [inline blocks/invoke.py:142]: not imported too

**Review by freedeaths — CHANGES_REQUESTED** [review]

File: database.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucklove) [reply database.py:47]: I've add a rule to forbid deleting active version: https://github.com/pingcap/LinguFlow/pull/1/files#diff-7336f211be5f6f21ce69bf332e0d14015b72ed0a8af6d15d8c0c98d40bf7b52eR400

**Review by freedeaths — APPROVED** [review]

**Pull request merged**